### ISSUE TITLE
Three functions for vertico that work from a candidate name rather than an index

### ISSUE THREAD

**Issue by rswgnu** [minad/vertico#379]
Hi Daniel:

I think these make it easier to control vertico programatically as they drive it based on candidates rather than having to know their indices.  I hope you will add them.

```elisp
(defun vertico-goto-candidate (candidate)
  "Go to an exact matching CANDIDATE and return its candidate index, else return nil."
  (when (active-minibuffer-window)
    (let ((index (cl-position candidate vertico--candidates :test 'equal)))
      (when index
	(with-selected-window (active-minibuffer-window)
	  (vertico--goto index)
	  index)))))

(defun vertico-insert-candidate (candidate)
  "Insert into minibuffer an exact matching CANDIDATE and return t, else return nil."
  (when (and (active-minibuffer-window) (vertico-goto-candidate candidate))
    (with-selected-window (active-minibuffer-window)
      (vertico-insert)
      t)))

 (defun vertico-exit-with-candidate (candidate)
   "Exit minibuffer utilizing an exact matching CANDIDATE and return t, else return nil."
   (when (and (active-minibuffer-window) (vertico-goto-candidate candidate))
    (with-selected-window (active-minibuffer-window)
      (vertico-exit)
      t)))
```

**Comment by minad** [comment]
Thanks, Robert! I think the [Vertico wiki](/minad/vertico/wiki) is the better place for the functions than the Vertico package, since these functions may be helpful but are not needed by Vertico itself. The functions are simple enough to maintain them inside another package or your user configuration. Since Vertico aims to be small and simple a programmatic API is not part of Vertico by design. I also wonder what do you use them for. How do you get access to a candidate string but not the corresponding index?

**Comment by rswgnu** [comment]
I am doing some work interfacing the context-sensitive Hyperbole Action Mouse Key to Vertico, separate from the Vertico Mouse extension library.  We wrap many separate systems in Hyperbole to produce a uniform experience.Hyperbole can grab the text of a completion without it having an index text property and then can use the functions provided to control Vertico the same way it does other completion systems.  No Vertico behavior change is necessary and Hyperbole mouse users can adopt Vertico with no behavior change on their end as well. I would rather have these basic functions that manipulate the index in Vertico than have to add them to Hyperbole, as it seems like a valuable, low maintenence way to offer this capability to any developers looking to control Vertico in simple ways.On Jul 4, 2023, at 4:54 AM, Daniel Mendler ***@***.***> wrote:﻿
Thanks, Robert! I think the Vertico wiki is the better place for the functions than the Vertico package, since these functions may be helpful but are not needed by Vertico itself. The functions are simple enough to maintain them inside another package or your user configuration. Since Vertico aims to be small and simple a programmatic API is not part of Vertico by design. I also wonder what do you use them for. How do you get access to a candidate string but not the corresponding index?

**Comment by minad** [comment]
> Hyperbole can grab the text of a completion without it having an index text property and then can use the functions provided to control Vertico the same way it does other completion systems.

This approach won't work properly since the displayed candidates do not necessarily correspond exactly to the actual candidate strings due to the way candidates are formatted and possibly truncated by Vertico. You must go via the candidate index as implemented in the vertico-mouse extension.

**Comment by rswgnu** [comment]
Ok.  Let me see if I understand.  If a user does not enable the mouse
extension, then no index text properties are assigned to the buffer text
lines and the main way Vertico has to determine the buffer text to index
correspondence is its current index pointer into an in-memory array of
actual completion strings, with the pointer updated each time a Vertico
movement command is issued.  This is why when you add mouse handling, you
add an index property to each buffer line to create the mapping that random
selection of a line requires.  Is that the gist of it?

The issue then is the need for a type of autoloading of the Vertico mouse
handling library when a mouse event occurs within a Vertico buffer.  Just
as you can't know when a command will be used and thus we have autoloads,
you can't know if the user has not yet enabled a Vertico extension but then
triggers an event that requires that extension.  But if you load the mouse
extension library after an external mouse event occurs, will it annotate
the existing displayed entries with the text properties so that the index
can be retrieved as part of processing the event that occurred before the
library was loaded (triggered by a mouse binding external to Vertico)?  Or
do we have to force loading of the mouse extension before we know whether
the user will ever use it or not?

**Comment by minad** [comment]
Yes, that's the gist. There is no nice solution for the autoloading problem except requiring that `vertico-mouse-mode` mode is enabled together with `vertico-mode`. However note that `vertico-mouse-mode` also makes sure that the candidates are highlighted on mouse hover. This feature already requires `vertico-mouse-mode` to be enabled early on.

**Comment by rswgnu** [comment]
Ok, I'll just work with that and pretty much have it integrated with
Hyperbole this way already.  As you say, the hover highlighting requires
early loading but if you load it later, programmatically things seem to
work the way I need them.  Thanks.
